godot-core/src/builtin/meta/mod.rs

```diff
@@ -7,27 +7,120 @@
 pub mod registration;
 
 mod class_name;
+mod godot_compat;
 mod return_marshal;
 mod signature;
 
 pub use class_name::*;
+pub use godot_compat::*;
 #[doc(hidden)]
 pub use return_marshal::*;
 #[doc(hidden)]
 pub use signature::*;
 
 use godot_ffi as sys;
+use sys::{GodotFfi, GodotNullableFfi};
 
 use crate::builtin::*;
 use crate::engine::global;
 use registration::method::MethodParamOrReturnInfo;
 
-/// Stores meta-information about registered types or properties.
+/// Conversion of GodotFfi-types into/from [`Variant`].
+pub trait GodotFfiVariant: Sized + GodotFfi {
+    fn ffi_to_variant(&self) -> Variant;
+    fn ffi_from_variant(variant: &Variant) -> Result<Self, VariantConversionError>;
+}
+
+mod sealed {
+    // To ensure the user does not implement `GodotType` for their own types.
+
+    use godot_ffi::GodotNullableFfi;
+
+    use super::GodotType;
+    use crate::builtin::*;
+    use crate::obj::*;
+
+    pub trait Sealed {}
+
+    impl Sealed for Aabb {}
+    impl Sealed for Basis {}
+    impl Sealed for Callable {}
+    impl Sealed for Vector2 {}
+    impl Sealed for Vector3 {}
+    impl Sealed for Vector4 {}
+    impl Sealed for Vector2i {}
+    impl Sealed for Vector3i {}
+    impl Sealed for Vector4i {}
+    impl Sealed for Quaternion {}
+    impl Sealed for Color {}
+    impl Sealed for GodotString {}
+    impl Sealed for StringName {}
+    impl Sealed for NodePath {}
+    impl Sealed for PackedByteArray {}
+    impl Sealed for PackedInt32Array {}
+    impl Sealed for PackedInt64Array {}
+    impl Sealed for PackedFloat32Array {}
+    impl Sealed for PackedFloat64Array {}
+    impl Sealed for PackedStringArray {}
+    impl Sealed for PackedVector2Array {}
+    impl Sealed for PackedVector3Array {}
+    impl Sealed for PackedColorArray {}
+    impl Sealed for Plane {}
+    impl Sealed for Projection {}
+    impl Sealed for Rid {}
+    impl Sealed for Rect2 {}
+    impl Sealed for Rect2i {}
+    impl Sealed for Signal {}
+    impl Sealed for Transform2D {}
+    impl Sealed for Transform3D {}
+    impl Sealed for Dictionary {}
+    impl Sealed for bool {}
+    impl Sealed for i64 {}
+    impl Sealed for i32 {}
+    impl Sealed for i16 {}
+    impl Sealed for i8 {}
+    impl Sealed for u64 {}
+    impl Sealed for u32 {}
+    impl Sealed for u16 {}
+    impl Sealed for u8 {}
+    impl Sealed for f64 {}
+    impl Sealed for f32 {}
+    impl Sealed for () {}
+    impl Sealed for Variant {}
+    impl<T: GodotType> Sealed for Array<T> {}
+    impl<T: GodotClass> Sealed for RawGd<T> {}
+    impl<T: GodotClass> Sealed for Gd<T> {}
```

Comment:
Do both need to be `GodotType`s here?

Reply:
No it isn't really possible, `RawGd` must implement `GodotType`. And unless we want people to use `RawGd` if they ever want to implement `GodotCompatible` with objects as the backing type, we'd need `Gd` to implement `GodotType`. And while i dont think people would want to use objects as the backing type often (it's likely easier to just make a `GodotClass` type), i dont see a particular reason to completely prevent people from doing so, and i'd rather people not use `RawGd` directly since it's meant to be a low-level type.

Reply:
Sounds good, let's leave it then 👍